### PR TITLE
Issue420

### PR DESCRIPTION
**Pull request by AmandaBirmingham**
While investigating the causes of #420 "Prepare Amplicon Sequencing Pool allows user to add metagenomics (not amplicon) plate pools, which leads to error downstream in sequencing run sample sheet preparation", I also found issues #497 "Pool list page checkboxes allows user to add metagenomics (not amplicon) plate pools, which leads to error downstream in sequencing run sample sheet preparation" (very similar to 420) and #498 "Prepare Sequencing Run allows users to mix amplicon and metagenomics pools, which leads to error downstream in sequencing run sample sheet preparation".  Basically all of these are related to the fact that although pools can be made up of any arbitrary contents (including any combination of amplicon library preps, shotgun library preps, and/or other pools), **it isn't currently sensical to *sequence* such a pool** as we don't know what the downstream outputs (sample sheets and/or prep sheets) for arbitrary pools (made up of mixtures of types of inputs) should look like.

For now, I solved all three of these issues by limiting which kinds of inputs are available to the user when creating an amplicon sequencing pool (can only input amplicon plate pools), a shotgun sequencing run (can only input shotgun plate pools), or an amplicon sequencing run (can only input amplicon sequencing pools).  In the future, we may decide that (a) there are realistic use-cases for mixing the types of inputs that go into a pool AND (b) we understand what the downstream outputs for sequencing runs of arbitrary pools should look like.  If and when that day comes, we will want to take out the limitations I have added here.

Note that these limitations *prevent* the case described in issue #444 "Modify generate_prep_information to correctly create prep sheets for sequencing runs based on library plate pools" from occurring, thus making it a low priority to fix.

Also note that there are some known code smells here ... for example, an examination of what kinds of library prep components go into a pool is now happening in `util.py`'s `get_pool_listing` as well as in `process.py`'s `SequencingProcess.create()` (the latter of which I am starting to think is the wrong place for it and is related to the bug for issue #498), and the definition of "amplicon_sequencing" pools as given in `PoolListHandler.get()` would also catch additional kinds of pools if we add additional protocols.  Fixing these smells will take a bit of thoughtful refactoring of how and where we identify different assay types; if this PR gets approved, I'll add an issue for that.

### CODE REVIEW

**Comment by AmandaBirmingham** [comment]
Because this PR includes various front-end fixes that are difficult to grok from the code alone, I'm adding comments below about what each of those fixes look like.

**Comment by AmandaBirmingham** [comment]
#420 "Prepare Amplicon Sequencing Pool allows user to add metagenomics (not amplicon) plate pools, which leads to error downstream in sequencing run sample sheet preparation"

Before PR:

Screen Shot 2019-05-06 at 9 33 56 PM

Screen Shot 2019-05-06 at 9 34 02 PM


After PR:

Screen Shot 2019-05-06 at 3 27 08 PM

Note that header clarifies this page is only for making an *amplicon* sequencing pool

Screen Shot 2019-05-06 at 3 27 16 PM

Note that only amplicon plate pools are available as inputs

**Comment by AmandaBirmingham** [comment]
#497 "Pool list page checkboxes allows user to add metagenomics (not amplicon) plate pools, which leads to error downstream in sequencing run sample sheet preparation" 

Before PR:

Screen Shot 2019-05-06 at 9 41 16 PM

Screen Shot 2019-05-06 at 9 41 29 PM

After PR:

Screen Shot 2019-05-06 at 3 25 39 PM

Note that pools that are NOT amplicon plate pools have grayed-out checkboxes because they cannot "legally" be used to make amplicon sequencing pools.

Screen Shot 2019-05-06 at 3 25 54 PM

Note that button has been relabeled to "Prepare amplicon sequencing pool" to match change to header on the corresponding page.

**Comment by AmandaBirmingham** [comment]
#498 "Prepare Sequencing Run allows users to mix amplicon and metagenomics pools, which leads to error downstream in sequencing run sample sheet preparation".

Before PR:

Screen Shot 2019-05-06 at 9 45 18 PM

Screen Shot 2019-05-06 at 9 45 57 PM

Screen Shot 2019-05-06 at 9 46 12 PM

After PR: 

Screen Shot 2019-05-06 at 10 03 51 PM

Now there are separate menu items for making a run from amplicon pools vs making a run from shotgun pools.  I went with this because I think it will be clear to the user, it can be supported with essentially no code duplication (just different input variables into the same html template page), and because having one single page *switch back and forth* between shotgun and amplicon input options and clear out state between them (a la `library_pooling.html`) has proven to be an invitation to bugs.

Screen Shot 2019-05-06 at 3 27 50 PM

Screen Shot 2019-05-06 at 3 28 02 PM

Note that if you pick the "prepare amplicon sequencing run" menu option, the header indicates this page is for amplicon inputs only and the pools you can choose from are ONLY the amplicon sequencing pools (so issue #444 can no longer be caused). Conversely, note that if you pick the "prepare shotgun sequencing run" menu option, the header indicates this page is for shotgun inputs only and the pools you can choose from are ONLY the shotgun plate pools

Screen Shot 2019-05-06 at 10 04 23 PM
 
Screen Shot 2019-05-06 at 3 28 24 PM

**Comment by charles-cowart** [comment]
> Note that pools that are NOT amplicon plate pools have grayed-out checkboxes because they cannot "legally" be used to make amplicon sequencing pools.

Would it be more clear to remove the non-amplicon plate pools?

**Comment by charles-cowart** [comment]
> and because having one single page switch back and forth between shotgun and amplicon input options and clear out state between them (a la library_pooling.html) has proven to be an invitation to bugs.
+1

**Comment by AmandaBirmingham** [comment]
> > Note that pools that are NOT amplicon plate pools have grayed-out checkboxes because they cannot "legally" be used to make amplicon sequencing pools.
> 
> Would it be more clear to remove the non-amplicon plate pools?

No, because this is occurring on the pool list page, which exists primarily for the purpose of allowing the user can see all the pools (of all kinds) that have been made.  Allowing the user to click on checkboxes next to them and shortcut amplicon sequencing pool creation is an add-on rather than the main goal (an add-on that only applies to certain kinds of pools :( ). 
 
There are a couple of other options besides graying out the checkboxes … we could make the user pick a pool type (as we make them pick the plate type in the plate list page), but we don’t currently have explicit pool types the way we have explicit plate types and that whole “switching types on the same page and clearing state” thing hasn’t worked out well for us in the past.  Alternately, we could entirely remove the checkboxes for the pools that can’t be included in amplicon sequencing pools, but I thought that would be MORE confusing to the user than graying them out (perhaps others disagree …).

File: labcontrol/gui/test/test_pool.py [hunk 1]
```diff
@@ -13,20 +13,65 @@
 
 
 class TestPoolHandlers(TestHandlerBase):
-    def test_get_plate_list_handler(self):
-        response = self.get('/pool_list')
+    def test_get_pool_list_handler_all(self):
+        response = self.get('/pool_list/all/')
```
Comment (charles-cowart) [inline labcontrol/gui/test/test_pool.py]: The '/all/' component of the resource string seems redundant - '/pool_list/' by itself, and with no further qualifying resource should display all resources by default, IMHO.

**Comment by charles-cowart** [comment]
Yeah, picking a pool type doesn't seem like a good route. What about graying out the text too? That might be less confusing than just graying out the box, but that's just my thinking.

File: labcontrol/gui/test/test_pool.py (hunk 1 above)
Comment (wasade) [reply labcontrol/gui/test/test_pool.py]: On the other hand, the internal behavior of the handler would be implicit. I'd be more in favor of explicit behavior

**Comment by wasade** [comment]
This is a thorough PR, and the additional contextual comments in the code are very beneficial. I'm strongly in favor of the explicitness of the approach, and the conservative behavior on handling pools. And I agree that the code smells can be deferred to a subsequent issue

**Comment by AmandaBirmingham** [comment]
Thanks, @wasade ! Do you have any opinions on the discussion above about the checkboxes? (see https://github.com/jdereus/labman/pull/500#issuecomment-489915946 and https://github.com/jdereus/labman/pull/500#issuecomment-489917864 ).  

I would prefer not to gray out the whole row for non-amplicon-plate pools, as I worry that would make the user think there was something wrong with them (and the user might well be on this page to do something completely unrelated to making amplicon sequencing pools from amplicon plate pools).  Also, more pragmatically, it would require me to make more javascript changes :D   

However, I do acknowledge that the grayed out checkboxes are not a great UI approach, so if anyone has a (preferably low-effort :) third option, I'm open to it ...

File: labcontrol/gui/test/test_pool.py (hunk 1 above)
Comment (AmandaBirmingham) [reply labcontrol/gui/test/test_pool.py]: @charles-cowart Given that @wasade likes the explicit behavior--and given that I actually don't know how to tweak the tornado handler regex to do both the explicit behavior and the implicit behavior in the same handler set-up :( -- would you be willing to table this idea?

**Comment by wasade** [comment]
> > > Note that pools that are NOT amplicon plate pools have grayed-out checkboxes because they cannot "legally" be used to make amplicon sequencing pools.
> > 
> > 
> > Would it be more clear to remove the non-amplicon plate pools?
> 
> No, because this is occurring on the pool list page, which exists primarily for the purpose of allowing the user can see all the pools (of all kinds) that have been made. Allowing the user to click on checkboxes next to them and shortcut amplicon sequencing pool creation is an add-on rather than the main goal (an add-on that only applies to certain kinds of pools :( ).

I agree here, while also agreeing that greying out the checkbox is not the most ideal as it can sometimes be difficult to differentiate.

Would it be complex to replace the greyed out checkbox with "N/A" ? One other option would be to add an italicized message at the bottom of the modal stating "_Greyed out check boxes indicate a pool that is not applicable (e.g., amplicon with shotgun)_". (or if "N/A" is preferable, making the italicized statement to that effect).

**Comment by AmandaBirmingham** [comment]
Based on https://github.com/jdereus/labman/pull/500#issuecomment-490249289, here's what the pool list page looks like now.  Please indicate if this works for y'all ... :)

Screen Shot 2019-05-07 at 1 58 43 PM

**Comment by AmandaBirmingham** [comment]
Note that I've also merged in the small change to master that was made after this PR was created.

File: labcontrol/gui/test/test_pool.py (hunk 1 above)
Comment (charles-cowart) [reply labcontrol/gui/test/test_pool.py]: definitely - not a problem.

**Pull request merged**